### PR TITLE
[SPARK-47850][SQL] Support `spark.sql.hive.convertInsertingUnpartitionedTable`

### PR DESCRIPTION
**Pull request by pan3793**
### What changes were proposed in this pull request?

This PR introduced a new configuration `spark.sql.hive.convertInsertingUnpartitionedTable` alongside the existing `spark.sql.hive.convertInsertingPartitionedTable` to allow fine grain switching from Hive Serde to Data Source on inserting Parquet/ORC Hive tables.

### Why are the changes needed?

In the practice of hybrid workload (Hive tables may be read/written by Hive, Spark, Impala, etc.), we usually use DataSource for reading Parquet/ORC tables but Hive Serde for writing, the current configuration combination allows us to achieve that except for unpartitioned tables.

### Does this PR introduce _any_ user-facing change?

No. The new added configuration `spark.sql.hive.convertInsertingUnpartitionedTable` default value is `true`, which keeps the existing behavior.

### How was this patch tested?

New UT is added.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Comment by pan3793** [comment]
cc @wangyum @yaooqinn @LuciferYang

**Review by LuciferYang — APPROVED** [review]
+1, LGTM (pending tests)

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveUtils.scala [hunk 1]
```diff
@@ -154,6 +154,16 @@ private[spark] object HiveUtils extends Logging {
...
 
+  val CONVERT_INSERTING_UNPARTITIONED_TABLE =
+    buildConf("spark.sql.hive.convertInsertingUnpartitionedTable")
+      .doc("When set to true, and `spark.sql.hive.convertMetastoreParquet` or " +
+        "`spark.sql.hive.convertMetastoreOrc` is true, the built-in ORC/Parquet writer is used" +
+        "to process inserting into unpartitioned ORC/Parquet tables created by using the HiveSQL " +
+        "syntax.")
+      .version("4.0.0")
+      .booleanConf
+      .createWithDefault(true)
```
Comment (yaooqinn) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveUtils.scala:165]: This changes the default behavior,right?

Comment (LuciferYang) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveUtils.scala:165]: seems unchanged

Comment (pan3793) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveUtils.scala:165]: No.

Previously, Spark always converts unpartitioned Hive tables to DataSource approach for writing when `spark.sql.hive.convertMetastore[Orc|Parquet]` is `true`.

**Review by dongjoon-hyun — COMMENTED** [review]
Please revisit the misleading PR title, @pan3793 .

Apache Spark already supports it, doesn't it?
> Support converting insert for unpartitioned Hive table

I believe "Support `spark.sql.hive.convertInsertingUnpartitionedTable`" might be more clear for the scope of this PR.

**Review by dongjoon-hyun — DISMISSED** [review]
Only one comment about the PR title.

**Review by yaooqinn — APPROVED** [review]

**Comment by pan3793** [comment]
@LuciferYang @yaooqinn @dongjoon-hyun Thanks for reviewing and approval, unfortunately, I found the converted plan has the same pattern as the Hive table CTAS case, which results in "Once strategy's idempotence is broken for batch Post-Hoc Resolution" when `spark.sql.hive.convertInsertingUnpartitionedTable=false`, we need a flag to distinguish the origin of `InsertIntoHiveTable` between Conversion and CTAS, https://github.com/apache/spark/pull/46052/commits/4dea7d0d9bcf4f4042900cce2958059641b2f36d introduced a TreeNodeTag to achieve that.

**Comment by pan3793** [comment]
Another approach I can imagine is, changing `Batch("Post-Hoc Resolution", Once` to `Batch("Post-Hoc Resolution", Fixed(1)` to skip idempotence check

**Comment by dongjoon-hyun** [comment]
Oh..

**Comment by pan3793** [comment]
cc @ulysses-you who made refactor on this part

**Review by ulysses-you — APPROVED** [review]

**Review by LuciferYang — APPROVED** [review]

**Comment by LuciferYang** [comment]
friendly ping @dongjoon-hyun, I would like to confirm if you have any other suggestions for this pr, thanks ~

**Comment by dongjoon-hyun** [comment]
Sorry guys. For now, I don't have any suggestion.

**Comment by LuciferYang** [comment]
Thanks for your feedback @dongjoon-hyun ~

**Comment by LuciferYang** [comment]
Merged into master for Spark 4.0. Thanks @pan3793 @yaooqinn @ulysses-you @dongjoon-hyun  ~

**Comment by pan3793** [comment]
@LuciferYang thank you and all

**Pull request closed without merging**